### DISCUSSION TITLE
firebase v9 not work in dev mode (No Firebase App '[DEFAULT]')

### DISCUSSION THREAD

**Discussion by niklv** [nuxt/nuxt#17924] [Questions]
### Environment

Nuxt CLI v3.0.0-27415326.3c563fa
RootDir: /Users/niklvov/Projects/fj_web_boilerplate
Nuxt project info:

------------------------------
- Operating System: `Darwin`
- Node Version:     `v16.14.0`
- Nuxt Version:     `3.0.0-27415326.3c563fa`
- Package Manager:  `npm@8.3.1`
- Bundler:          `Vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

I want to create a firebase plugin that provides auth module

`npm i firebase`

```typescript
// nuxt.config.ts
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({})


// plugins/firebase.ts
import { initializeApp } from 'firebase/app'
import { getAuth } from 'firebase/auth'

export default defineNuxtPlugin(() => {
  // if you have a firebase project you can pass your credentials, for minimal reproduction see notes below
  initializeApp({ }) 
  return {
    provide: {
      auth: getAuth()
    }
  }
})
```

Then i run `npx nuxi dev`, and open `http://localhost:3000`.
In server console I get this error:

```
Nuxt CLI v3.0.0-27415326.3c563fa

> Local:    http://localhost:3000/
> Network:  http://192.168.1.121:3000/

ℹ Vite warmed up in 838ms
ℹ Vite warmed up in 2272ms
✔ Vite server built in 5550ms
✔ Nitro built in 780 ms

Firebase: No Firebase App '[DEFAULT]' has been created - call Firebase App.initializeApp() (app/no-app).
  at getApp (file://./node_modules/@firebase/app/dist/esm/index.esm2017.js:459:29)
  at Module.getAuth (file://./node_modules/@firebase/auth/dist/node-esm/index-2bbaab7c.js:5980:24)
  at file://./.nuxt/dist/server/server.mjs:2170:35
  at callWithNuxt (file://./.nuxt/dist/server/server.mjs:262:20)
  at applyPlugin (file://./.nuxt/dist/server/server.mjs:213:29)
  at Module.applyPlugins (file://./.nuxt/dist/server/server.mjs:223:11)
  at async createNuxtAppServer (file://./.nuxt/dist/server/server.mjs:51:5)
  at async renderToString (file://./node_modules/vue-bundle-renderer/dist/index.mjs:247:19)
  at async renderMiddleware (file://./.nuxt/nitro/index.mjs:206:20)
  at async handle (file://./node_modules/h3/dist/index.mjs:601:19)
```
Even if i pass corrent config into `initializeApp` if fails to with same error.

But if I run `npx nuxi build && npx nuxi preview` and open `http://localhost:3000`.
In server console I get different error:
```
Nuxt CLI v3.0.0-27415326.3c563fa

ℹ Node.js version: 16.14.0
ℹ Preset: server
ℹ Working dir: .output
ℹ Starting preview command: node ./server/index.mjs
Listening on http://localhost:3000/

Firebase: Error (auth/invalid-api-key).
  at createErrorInternal (file://./server/node_modules/@firebase/auth/dist/node-esm/index-2bbaab7c.js:466:40)  
  at _assert (file://./server/node_modules/@firebase/auth/dist/node-esm/index-2bbaab7c.js:470:15)  
  at file://./server/node_modules/@firebase/auth/dist/node-esm/index-2bbaab7c.js:5919:13  
  at Component.instanceFactory (file://./server/node_modules/@firebase/auth/dist/node-esm/index-2bbaab7c.js:5936:11)  
  at Provider.getOrInitializeService (file://./server/node_modules/@firebase/component/dist/esm/index.esm2017.js:290:39)  
  at Provider.initialize (file://./server/node_modules/@firebase/component/dist/esm/index.esm2017.js:234:31)  
  at initializeAuth (file://./server/node_modules/@firebase/auth/dist/node-esm/index-2bbaab7c.js:584:27)  
  at getAuth (file://./server/node_modules/@firebase/auth/dist/node-esm/index-2bbaab7c.js:5985:12)  
  at file://./server/chunks/app/server.mjs:3254:13  
  at callWithNuxt (file://./server/chunks/app/server.mjs:335:20)
```

Error is different.
And if i pass correct config into `initializeApp` it did not fails at all. 
This marks that production build code is working properly and dev code not.

### Describe the bug

Nuxt3 fails in dev mode with firebase-js-sdk module.
Production build is works as expected.

**Comment by Diizzayy** [comment]
This issue can be closed

@niklv This occurs because your plugin attempts to initialize firebase in server side, where as it should only occur in browser.

Your firebase plugin needs to be changed to a client-only plugin.

`nuxt.config.ts`
```js
import { defineNuxtConfig } from "nuxt3";

export default defineNuxtConfig({
  publicRuntimeConfig: {
    firebaseConfig: {},
  },
});
```

`plugins/firebase.client.ts`: this is a working code snippet from my personal project
```js
import { getApps, initializeApp } from "firebase/app";
import { getAuth, setPersistence, inMemoryPersistence } from "firebase/auth";

export default defineNuxtPlugin(() => {
  try {
    if (getApps().length === 0) {
      const { firebaseConfig } = useRuntimeConfig();
      if (firebaseConfig) initializeApp(firebaseConfig);
    }

    const auth = getAuth();
    if (auth) setPersistence(auth, inMemoryPersistence);
  } catch (e) {
    console.error(e);
  }
});
```

also, with firebase v9. I think it's very pointless to provide / inject anything from firebase

**Comment by niklv** [comment]
@Diizzayy I what to use firebase-js-sdk in ssr to. So i what to initialize in server and client plugin. (And I named it `firebase.ts` especially).
Answer not covered issue that is initialize ok in production build and no initialized in development build.
@danielroe please reopen

**Reply by Diizzayy** [reply]
Firebase-js-sdk cannot be used in SSR, it's client-side only. You have to utilize firebase-admin for any server side functionality.

**Reply by niklv** [reply]
No, you can user firebase-js-sdk in SSR. Please check older nuxt module https://github.com/nuxt-community/firebase-module
I plan user firebase-admin to validate input token later, but it is outside my issue.

**Reply by Diizzayy** [reply]
The module you referenced actually uses firebase-admin internally for SSR functionality.

**Reply by niklv** [reply]
Yes, please check that firebase-admin do in this module. All request to firestore works via firebase-js-sdk in ssr.

**Reply by Diizzayy** [reply]
This is something that I myself have previously worked on in the past.

In order to use Firestore in SSR, it must be handled via `firebase-admin`. The `firebase-js-sdk` is literally only for browser environments.

Your code must take account of the running environment and utilize either `firebase-admin` or the `firebase-js-sdk` accordingly.

@niklv I'd be happy to help you put together a minimal implementation if you'd like

**Reply by niklv** [reply]
@Diizzayy Implementation in SSR via firebase-admin will not assume working security rules (firestore/rtdb).
My plan:
Use `firebase-js-sdk` in SSR to make requests to firestore.
`firebase-admin` in SSR will check idToken in request and somehow inject it into firebase-js-sdk app (probably with using some non public methods/fields or js Proxy).

As you can see there https://firebase.google.com/docs/web/environments-js-sdk `firebase-js-sdk` should work in node enviroment. So it should work in SSR.

I'll appreciate any help in this process. Thank you.

**Comment by danielroe** [comment]
I'll convert this to a discussion 👍

**Reply by niklv** [reply]
Sorry, but it is a bug. Same code get different behaviours in `npx nuxi dev` vs `npx nuxi build && npx nuxi preview`.

**Reply by danielroe** [reply]
If you are experiencing a bug, would you create a new issue with a reproduction please? 🙏  (I need to see the entire code, particularly as the error you're getting is about an API key which it's saying you're not providing.)

**Reply by niklv** [reply]
Ok

**Reply by niklv** [reply]
@danielroe I created repo with reproduction code https://github.com/niklv/nuxt3_firebase9_ssr_bug and issue #3294

**Reply by niklv** [reply]
@danielroe if you need credentials to my firebase poject, please mail me niklvov@gmail.com

**Comment by wobsoriano** [comment]
I have a starter you might be interested

https://github.com/wobsoriano/nuxt-firebase-auth-starter
